File: Design-Patterns-Using-Swift/AbstractFactory/AbstractFactoryDemo.swift
// Menu protocol
protocol Menu {
    func desc()
}

// WinMenu class
class WinMenu: Menu {
    func desc() {
        print("Win Menu!!")
    }
}

// MacMenu class
class MacMenu: Menu {
    func desc() {
        print("Mac Menu!!")
    }
}

// Button protocol
protocol Button {
    func desc()
}

// WinButton class
class WinButton: Button {
    func desc() {
        print("Win Button!!")
    }
}

// MacButton class
class MacButton: Button {
    func desc() {
        print("Mac Button!!")
    }
}

// Abstract Factory protocol
protocol AbstractFactory {
    func getMenu() -> Menu
    func getButton() -> Button
}

// WinFactory class
class WinFactory: AbstractFactory {
    func getMenu() -> Menu {
        return WinMenu()
    }

    func getButton() -> Button {
        return WinButton()
    }
}

// MacFactory class
class MacFactory: AbstractFactory {
    func getMenu() -> Menu {
        return MacMenu()
    }

    func getButton() -> Button {
        return MacButton()
    }
}

// Client code
let macFactory: AbstractFactory = MacFactory()
macFactory.getMenu().desc()
macFactory.getButton().desc()

let winFactory: AbstractFactory = WinFactory()
winFactory.getButton().desc()
winFactory.getMenu().desc()

/*
Mac Menu!!
Mac Button!!
Win Button!!
Win Menu!!
*/

File: Design-Patterns-Using-Kotlin/AbstractFactory/AbstractFactoryDemo.kt
// Menu interface
interface Menu {
    fun desc()
}

// WinMenu class
class WinMenu : Menu {
    override fun desc() {
        println("Win Menu!!")
    }
}

// MacMenu class
class MacMenu : Menu {
    override fun desc() {
        println("Mac Menu!!")
    }
}

// Button interface
interface Button {
    fun desc()
}

// WinButton class
class WinButton : Button {
    override fun desc() {
        println("Win Button!!")
    }
}

// MacButton class
class MacButton : Button {
    override fun desc() {
        println("Mac Button!!")
    }
}

// Abstract Factory interface
interface AbstractFactory {
    val menu: Menu
    val button: Button
}

// WinFactory class
class WinFactory : AbstractFactory {
    override val menu: Menu
        get() = WinMenu()

    override val button: Button
        get() = WinButton()
}

// MacFactory class
class MacFactory : AbstractFactory {
    override val menu: Menu
        get() = MacMenu()

    override val button: Button
        get() = MacButton()
}

// Client code
fun main() {
    val macFactory: AbstractFactory = MacFactory()
    macFactory.menu.desc()
    macFactory.button.desc()

    val winFactory: AbstractFactory = WinFactory()
    winFactory.button.desc()
    winFactory.menu.desc()
}

/*
Mac Menu!!
Mac Button!!
Win Button!!
Win Menu!!
*/